classes/object_file_system.php

```diff
@@ -411,4 +411,76 @@ public function get_maximum_upload_filesize() {
         return $this->externalclient->get_maximum_upload_size();
     }
 
+    /**
+     * Extends remove_file to delete file from remote object storage.
+     *
+     * @param string $contenthash
+     */
+    public function remove_file($contenthash) {
+        if (!self::is_file_removable($contenthash)) {
+            // Don't remove the file - it's still in use.
+            return;
+        }
+
+        $location = $this->get_object_location_from_hash($contenthash);
+        if ($location == OBJECT_LOCATION_ERROR) {
+            return;
+        }
+
+        $trashpath  = $this->get_trash_fulldir_from_hash($contenthash);
+        $trashfile  = $this->get_trash_fullpath_from_hash($contenthash);
+
+        if (!is_dir($trashpath)) {
+            mkdir($trashpath, $this->dirpermissions, true);
+        }
+
+        if (file_exists($trashfile)) {
+            // A copy of this file is already in the trash.
+            // Remove the old version.
+            $this->delete_object_from_hash($contenthash);
+            return;
+        }
+
+        // Move the contentfile to the trash, and fix permissions as required.
+        $this->copy_file_from_hash_to_path($contenthash, $trashfile);
+        $this->delete_object_from_hash($contenthash);
+
+        // Fix permissions, only if needed.
+        $currentperms = octdec(substr(decoct(fileperms($trashfile)), -4));
+        if ((int)$this->filepermissions !== $currentperms) {
+            chmod($trashfile, $this->filepermissions);
+        }
+    }
+
+    public function copy_file_from_hash_to_path($contenthash, $destinationpath) {
+        $path = $this->get_remote_path_from_hash($contenthash);
+        copy ($path, $destinationpath);
+    }
+
+    public function delete_object_from_hash($contenthash) {
+        $location = $this->get_object_location_from_hash($contenthash);
+
+        $config = get_objectfs_config();
+        switch ($location) {
+            case OBJECT_LOCATION_LOCAL:
+                $path = $this->get_local_path_from_hash($contenthash);
+                unlink($path);
+                break;
+            case OBJECT_LOCATION_DUPLICATED:
+                $path = $this->get_local_path_from_hash($contenthash);
```

Comment:
Refactor this into a sub method delete_local_file_from_hash. Reuse above to reduce duplication.